File: routes/sync.js
const express = require('express');
const router = express.Router();
const SyncService = require('../services/syncService');
const serviceRegistry = require('../services/serviceRegistry');

/**
 * Check if required services are connected for the sync operation
 */
const requireServices = (req, res, next) => {
  const { sourceType, destinationType } = req.body;
  
  if (!sourceType || !destinationType) {
    return res.status(400).json({ 
      error: 'Both sourceType and destinationType are required' 
    });
  }

  const validation = serviceRegistry.validateSyncRequirements(
    req.session, 
    sourceType, 
    destinationType
  );

  if (!validation.success) {
    return res.status(401).json({ error: validation.message });
  }

  next();
};

/**
 * Start a playlist sync operation between any two services
 */
router.post('/playlist', requireServices, async (req, res) => {
  const { 
    sourceType, 
    destinationType, 
    sourcePlaylistId, 
    options = {} 
  } = req.body;

  if (!sourcePlaylistId) {
    return res.status(400).json({ error: 'Source playlist ID required' });
  }

  res.writeHead(200, {
    'Content-Type': 'text/event-stream',
    'Cache-Control': 'no-cache',
    'Connection': 'keep-alive',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Cache-Control'
  });

  const sendEvent = (type, data) => {
    res.write(`event: ${type}\n`);
    res.write(`data: ${JSON.stringify(data)}\n\n`);
  };

  // Send heartbeat every 10 seconds to keep connection alive
  const heartbeat = setInterval(() => {
    sendEvent('heartbeat', { timestamp: Date.now() });
  }, 10000);

  try {
    // Create sync service from session data
    const syncService = SyncService.createFromSession(
      req.session, 
      sourceType, 
      destinationType
    );

    const syncOptions = {
      ...options,
      onProgress: (progress, current, total) => {
        sendEvent('progress', { 
          progress, 
          current, 
          total,
          sourceService: sourceType,
          destinationService: destinationType
        });
      }
    };

    sendEvent('start', { 
      message: `Starting sync from ${sourceType} to ${destinationType}...`,
      sourceService: sourceType,
      destinationService: destinationType
    });

    const result = await syncService.syncPlaylist(sourcePlaylistId, syncOptions);

    sendEvent('complete', {
      success: true,
      result: {
        sourceService: result.sourceService,
        destinationService: result.destinationService,
        playlistName: result.sourcePlaylist?.name,
        totalTracks: result.totalTracks,
        matchedTracks: result.matchedTracks,
        unmatchedTracks: result.unmatchedTracks,
        duration: result.duration,
        destinationPlaylistId: result.destinationPlaylist?.id,
        syncMode: result.syncMode
      }
    });

  } catch (error) {
    console.error('Sync error:', error);
    sendEvent('error', {
      success: false,
      error: error.message,
      sourceService: sourceType,
      destinationService: destinationType
    });
  } finally {
    clearInterval(heartbeat);
    res.end();
  }
});

/**
 * Validate a sync operation before starting
 */
router.post('/validate', async (req, res) => {
  try {
    const { sourceType, destinationType, sourcePlaylistId } = req.body;

    if (!sourceType || !destinationType || !sourcePlaylistId) {
      return res.status(400).json({ 
        error: 'sourceType, destinationType, and sourcePlaylistId are required' 
      });
    }

    // Validate service connections
    const validation = serviceRegistry.validateSyncRequirements(
      req.session, 
      sourceType, 
      destinationType
    );

    if (!validation.success) {
      return res.status(400).json({ 
        valid: false, 
        error: validation.message 
      });
    }

    // Create sync service and validate the specific playlist
    const syncService = SyncService.createFromSession(
      req.session, 
      sourceType, 
      destinationType
    );

    const playlistValidation = await syncService.validateSync(sourcePlaylistId);

    if (!playlistValidation.valid) {
      return res.status(400).json({
        valid: false,
        error: playlistValidation.error
      });
    }

    res.json({
      valid: true,
      sourcePlaylist: playlistValidation.sourcePlaylist,
      capabilities: playlistValidation.capabilities
    });

  } catch (error) {
    console.error('Validation error:', error);
    res.status(500).json({ 
      valid: false, 
      error: 'Failed to validate sync operation' 
    });
  }
});

/**
 * Check if a playlist already exists in the destination service
 */
router.post('/check-existing', async (req, res) => {
  try {
    const { sourceType, destinationType, sourcePlaylistId } = req.body;

    if (!sourceType || !destinationType || !sourcePlaylistId) {
      return res.status(400).json({ 
        error: 'sourceType, destinationType, and sourcePlaylistId are required' 
      });
    }

    const syncService = SyncService.createFromSession(
      req.session, 
      sourceType, 
      destinationType
    );
    
    // Get source playlist details
    const sourcePlaylist = await syncService.sourceService.getPlaylistDetails(sourcePlaylistId);
    
    // Check if it exists in destination service
    const existingPlaylist = await syncService.destinationService.findPlaylistByName(
      sourcePlaylist.name
    );
    
    let trackCount = 0;
    if (existingPlaylist) {
      const tracks = await syncService.destinationService.getPlaylistTracks(existingPlaylist.id);
      trackCount = tracks.length;
    }
    
    res.json({
      sourcePlaylist: {
        name: sourcePlaylist.name,
        trackCount: sourcePlaylist.trackCount
      },
      existingPlaylist: existingPlaylist ? {
        id: existingPlaylist.id,
        name: existingPlaylist.name,
        trackCount
      } : null,
      sourceService: sourceType,
      destinationService: destinationType
    });
  } catch (error) {
    console.error('Error checking existing playlist:', error);
    res.status(500).json({ error: 'Failed to check existing playlist' });
  }
});

/**
 * Get sync capabilities between two services
 */
router.post('/capabilities', (req, res) => {
  try {
    const { sourceType, destinationType } = req.body;

    if (!sourceType || !destinationType) {
      return res.status(400).json({ 
        error: 'sourceType and destinationType are required' 
      });
    }

    // Validate service connections
    const validation = serviceRegistry.validateSyncRequirements(
      req.session, 
      sourceType, 
      destinationType
    );

    if (!validation.success) {
      return res.status(400).json({ 
        error: validation.message 
      });
    }

    const syncService = SyncService.createFromSession(
      req.session, 
      sourceType, 
      destinationType
    );

    const capabilities = syncService.getSyncCapabilities();

    res.json(capabilities);

  } catch (error) {
    console.error('Error getting capabilities:', error);
    res.status(500).json({ error: 'Failed to get sync capabilities' });
  }
});

/**
 * Get overall sync status (which services are connected)
 */
router.get('/status', (req, res) => {
  try {
    const connectionStatus = serviceRegistry.getConnectionStatus(req.session);
    const connectedUsers = serviceRegistry.getConnectedServiceUsers(req.session);
    const availableServices = serviceRegistry.getAvailableServices();

    const services = availableServices.map(service => ({
      ...service,
      connected: connectionStatus[service.type] || false,
      user: connectedUsers[service.type] || null
    }));

    res.json({
      authenticated: Object.values(connectionStatus).some(Boolean),
      services,
      totalConnected: Object.values(connectionStatus).filter(Boolean).length
    });
  } catch (error) {
    console.error('Error getting sync status:', error);
    res.status(500).json({ error: 'Failed to get sync status' });
  }
});

module.exports = router;

File: app/routes/sync.py
from fastapi import APIRouter, Request, HTTPException
from fastapi.responses import StreamingResponse
import json
import asyncio
import time

from app.services.sync_service import SyncService
from app.services.service_registry import ServiceRegistry

router = APIRouter()

async def stream_sync_progress(sync_service: SyncService, source_playlist_id: str, options: dict):
    """Stream sync progress using Server-Sent Events"""

    def send_event(event_type: str, data: dict):
        """Format and return SSE event"""
        return f"event: {event_type}\ndata: {json.dumps(data)}\n\n"

    # Send initial event
    yield send_event("start", {
        "message": f"Starting sync from {sync_service.source_service.service_name} to {sync_service.destination_service.service_name}...",
        "sourceService": sync_service.source_service.service_name,
        "destinationService": sync_service.destination_service.service_name
    })

    # Progress tracking
    progress_data = {"current": 0, "total": 0, "progress": 0}

    def on_progress(progress: int, current: int, total: int):
        progress_data.update({
            "progress": progress,
            "current": current,
            "total": total,
            "sourceService": sync_service.source_service.service_name,
            "destinationService": sync_service.destination_service.service_name
        })

    # Add progress callback to options
    options["on_progress"] = lambda p, c, t: None  # We'll handle progress differently

    try:
        # Start sync
        result = await sync_service.sync_playlist(source_playlist_id, options)

        # Send completion event
        yield send_event("complete", {
            "success": True,
            "result": {
                "sourceService": result.source_service,
                "destinationService": result.destination_service,
                "playlistName": result.source_playlist.name if result.source_playlist else None,
                "totalTracks": result.total_tracks,
                "matchedTracks": result.matched_tracks,
                "unmatchedTracks": result.unmatched_tracks,
                "duration": result.duration,
                "destinationPlaylistId": result.destination_playlist.id if result.destination_playlist else None,
                "syncMode": result.sync_mode
            }
        })

    except Exception as error:
        print(f"Sync error: {error}")
        yield send_event("error", {
            "success": False,
            "error": str(error),
            "sourceService": sync_service.source_service.service_name,
            "destinationService": sync_service.destination_service.service_name
        })

@router.post("/playlist")
async def sync_playlist(request: Request):
    """Start a playlist sync operation between two services"""
    try:
        body = await request.json()
        source_type = body.get("sourceType")
        destination_type = body.get("destinationType")
        source_playlist_id = body.get("sourcePlaylistId")
        options = body.get("options", {})

        if not all([source_type, destination_type, source_playlist_id]):
            raise HTTPException(
                status_code=400,
                detail="sourceType, destinationType, and sourcePlaylistId are required"
            )

        # Validate service connections
        validation = ServiceRegistry.validate_sync_requirements(
            request.session, source_type, destination_type
        )

        if not validation["success"]:
            raise HTTPException(status_code=401, detail=validation["message"])

        # Create sync service
        sync_service = SyncService.create_from_session(
            request.session, source_type, destination_type
        )

        # Return streaming response
        return StreamingResponse(
            stream_sync_progress(sync_service, source_playlist_id, options),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "Cache-Control"
            }
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start sync: {str(e)}")

@router.post("/validate")
async def validate_sync(request: Request):
    """Validate a sync operation before starting"""
    try:
        body = await request.json()
        source_type = body.get("sourceType")
        destination_type = body.get("destinationType")
        source_playlist_id = body.get("sourcePlaylistId")

        if not all([source_type, destination_type, source_playlist_id]):
            raise HTTPException(
                status_code=400,
                detail="sourceType, destinationType, and sourcePlaylistId are required"
            )

        # Validate service connections
        validation = ServiceRegistry.validate_sync_requirements(
            request.session, source_type, destination_type
        )

        if not validation["success"]:
            return {"valid": False, "error": validation["message"]}

        # Create sync service and validate the specific playlist
        sync_service = SyncService.create_from_session(
            request.session, source_type, destination_type
        )

        playlist_validation = await sync_service.validate_sync(source_playlist_id)

        if not playlist_validation["valid"]:
            return {"valid": False, "error": playlist_validation["error"]}

        return {
            "valid": True,
            "sourcePlaylist": {
                "id": playlist_validation["source_playlist"].id,
                "name": playlist_validation["source_playlist"].name,
                "trackCount": playlist_validation["source_playlist"].track_count
            },
            "capabilities": playlist_validation["capabilities"]
        }

    except Exception as e:
        return {"valid": False, "error": "Failed to validate sync operation"}

@router.post("/check-existing")
async def check_existing_playlist(request: Request):
    """Check if a playlist already exists in the destination service"""
    try:
        body = await request.json()
        source_type = body.get("sourceType")
        destination_type = body.get("destinationType")
        source_playlist_id = body.get("sourcePlaylistId")

        if not all([source_type, destination_type, source_playlist_id]):
            raise HTTPException(
                status_code=400,
                detail="sourceType, destinationType, and sourcePlaylistId are required"
            )

        # Create sync service
        sync_service = SyncService.create_from_session(
            request.session, source_type, destination_type
        )

        # Get source playlist details
        source_playlist = sync_service.source_service.get_playlist_details(source_playlist_id)

        # Check if it exists in destination service
        existing_playlist = sync_service.destination_service.find_playlist_by_name(
            source_playlist.name
        )

        track_count = 0
        if existing_playlist:
            tracks = sync_service.destination_service.get_playlist_tracks(existing_playlist.id)
            track_count = len(tracks)

        return {
            "sourcePlaylist": {
                "name": source_playlist.name,
                "trackCount": source_playlist.track_count
            },
            "existingPlaylist": {
                "id": existing_playlist.id,
                "name": existing_playlist.name,
                "trackCount": track_count
            } if existing_playlist else None,
            "sourceService": source_type,
            "destinationService": destination_type
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to check existing playlist: {str(e)}")

@router.post("/capabilities")
async def get_sync_capabilities(request: Request):
    """Get sync capabilities between two services"""
    try:
        body = await request.json()
        source_type = body.get("sourceType")
        destination_type = body.get("destinationType")

        if not all([source_type, destination_type]):
            raise HTTPException(
                status_code=400,
                detail="sourceType and destinationType are required"
            )

        # Validate service connections
        validation = ServiceRegistry.validate_sync_requirements(
            request.session, source_type, destination_type
        )

        if not validation["success"]:
            raise HTTPException(status_code=400, detail=validation["message"])

        # Create sync service
        sync_service = SyncService.create_from_session(
            request.session, source_type, destination_type
        )

        capabilities = sync_service.get_sync_capabilities()
        return capabilities

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get sync capabilities: {str(e)}")

@router.get("/status")
async def get_sync_status(request: Request):
    """Get overall sync status (which services are connected)"""
    try:
        connection_status = ServiceRegistry.get_connection_status(request.session)
        connected_users = ServiceRegistry.get_connected_service_users(request.session)
        available_services = ServiceRegistry.get_available_services()

        services = []
        for service in available_services:
            services.append({
                **service,
                "connected": connection_status.get(service["type"], False),
                "user": connected_users.get(service["type"])
            })

        return {
            "authenticated": any(connection_status.values()),
            "services": services,
            "totalConnected": sum(1 for connected in connection_status.values() if connected)
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get sync status: {str(e)}")